### PR TITLE
[flang] Add debug information for module variables.

### PR DESCRIPTION
**Pull request by abidh**
This PR add debug info for module variables. The module variables are added as global variables but their scope is set to module instead of compile unit. The scope of function declared inside a module is also set accordingly.

After this patch, a module variable could be evaluated in the GDB as `p helper::gli` where helper is name of the module and gli is the name of the variable. A future patch will add the import module functionality which will remove the need to prefix the name with helper::.

The line number where is module is declared is a best guess at the moment as this information is not part of the GlobalOp.

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp [hunk 1]
```diff
@@ -1030,10 +1030,15 @@ LogicalResult ModuleTranslation::convertGlobals() {
       llvm::DIGlobalVariable *diGlobalVar = diGlobalExpr->getVariable();
       var->addDebugInfo(diGlobalExpr);
 
+      // For fortran, the scope hierarchy can be
+      // variable -> module -> compile unit
+      llvm::DIScope *scope = diGlobalVar->getScope();
+      if (llvm::DIModule *mod = dyn_cast_if_present<llvm::DIModule>(scope))
+        scope = mod->getScope();
+
```
Comment (kiranchandramohan) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1038]: Changes to MLIR should go in as a separate patch. And a test is required for the MLIR changes in I guess `mlir/test/Target/LLVMIR/llvmir-debug.mlir`.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 2]
```diff
@@ -48,10 +49,87 @@ class AddDebugInfoPass : public fir::impl::AddDebugInfoBase<AddDebugInfoPass> {
...
+  if (auto fileLoc = mlir::dyn_cast<mlir::FileLineColLoc>(loc))
+    line = fileLoc.getLine();
+  return line;
+}
+
 } // namespace
 
+// The `module` does not have a first class representation in the `FIR`. We
+// extract information about it from the name of the identifiers and keep a
+// map to avoid duplication.
```
Comment (kiranchandramohan) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:75]: Would this mean that if a module variable or subroutine is not used we will not have the information that the module is used via a USE statement?

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp (hunk 1 above)
Comment (abidh) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1038]: Thanks for letting me know. I have now opened #91604 with this change and a testcase. I will remove this bit from this PR.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 2 above)
Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:75]: In debugger, one should be able to evaluate module variables using module::var syntax even in functions where the module is not used (using `use` keyword). For functions where it is used, evaluation without module:: should work too (once we have implemented the imported module functionality). 

I also tried multi file examples where module is defined in one file and used in others. In such cases, compilers (e.g. gfortran, classic flang) generate a module entry in both the compile units. To differentiate, they set decl attribute to true in files where it is just used. I have tried to follow the similar approach. I found that GlobalOp::hasInitializationBody() gave me this information. It was true in compile unit where the module was actually defined and false where it was just used.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 3]
```diff
@@ -91,6 +169,10 @@ void AddDebugInfoPass::runOnOperation() {
       llvm::dwarf::getLanguage("DW_LANG_Fortran95"), fileAttr, producer,
       isOptimized, debugLevel);
 
+  module.walk([&](fir::GlobalOp globalOp) {
```
Comment (tblah) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:172]: `mlir::Operation::walk` will visit every descendant operation. I think `fir::GlobalOp` can only be nested directly inside of the `MoudleOp` so it would be much more efficient to have a for loop over the direct descendants of the module.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 4]
```diff
@@ -48,10 +49,87 @@ class AddDebugInfoPass : public fir::impl::AddDebugInfoBase<AddDebugInfoPass> {
 public:
   AddDebugInfoPass(fir::AddDebugInfoOptions options) : Base(options) {}
   void runOnOperation() override;
+
+private:
+  std::map<std::string, mlir::LLVM::DIModuleAttr> moduleMap;
```
[the remark this answers is not in the document]
Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp]: Done.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 3 above)
Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp]: I have changed it to
`for (auto globalOp : module.getOps<fir::GlobalOp>())`

Please let me know if that was not what you had in mind.

Comment (tblah) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp]: Yes that's what I meant. Thank you!

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 5]
```diff
@@ -34,6 +34,7 @@
 #include "llvm/Support/FileSystem.h"
 #include "llvm/Support/Path.h"
 #include "llvm/Support/raw_ostream.h"
+#include <map>
```
Comment (tblah) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:37]: nit: this is no longer needed

**Review by tblah — APPROVED** [review]
This looks good to me, thanks!

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 5 above)
Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp]: Thanks for catching this. Removed.

File: flang/test/Transforms/debug-module-1.f90 [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+  glr = 3.14
+  gli = 2
+  call test()
+
+end program test
+
+! CHECK-DAG: loc(fused<#[[GLR]]>[#[[LOC1]]])
+! CHECK-DAG: loc(fused<#[[GLI]]>[#[[LOC2]]])
+! CHECK-DAG: loc(fused<#[[TEST]]>[#[[LOC3]]])
```
Comment (kiranchandramohan) [inline flang/test/Transforms/debug-module-1.f90:39]: Nit: end of line.

**Review by kiranchandramohan — COMMENTED** [review]
The patch adds debug info for other kind of globals (besides module variables) like for eg. `x` in the following case. Is that the intention? If so a test would be good.
```
subroutine abc
  integer :: x = 2
  print *, x
end subroutine
```
Commonblocks also are globals but I did not see debug for them. 

I am guessing the following points will come in future patches.
-> Renamed module variables using `use` statement.
-> Private module variables. Can they be distinguished in DWARF?
-> Submodules

**Comment by kiranchandramohan** [comment]
There is a plan for a pass (https://github.com/llvm/llvm-project/pull/73829) that would convert some constants to globals. Would it be OK to mark this as a Global Variable?

**Comment by abidh** [comment]
> The patch adds debug info for other kind of globals (besides module variables) like for eg. `x` in the following case. Is that the intention? If so a test would be good.
> 
> ```
> subroutine abc
>   integer :: x = 2
>   print *, x
> end subroutine
> ```
> 
This was not intentional. The variable with save attribute should be added as global but with scope limited to the subroutine. I will add a check to skip them in this patch.

On a related note, while processing `DeclareOp` of such variables, are you aware of anything that will tell me that these variables have `save` attribute? I ask because I noticed that they are being added as local variable too.

> Commonblocks also are globals but I did not see debug for them.
> 

Yes, common blocks are not supported yet.

> I am guessing the following points will come in future patches. -> Renamed module variables using `use` statement. -> Private module variables. Can they be distinguished in DWARF? -> Submodules

Private module variable should work with this patch. I was not aware of the renaming by the `use` statement syntax. I will add it in my todo list. `DWARF` has concept of `DW_TAG_imported_declaration` for such cases. I will have to see if we have enough information to generate it.

**Comment by abidh** [comment]
> There is a plan for a pass (#73829) that would convert some constants to globals. Would it be OK to mark this as a Global Variable?

Currently, we add variable only if deconstructed name kind is `fir::NameUniquer::NameKind::VARIABLE`. I think that should filter out the generated globals.

File: flang/test/Transforms/debug-module-1.f90 (hunk 6 above)
Comment (abidh) [reply flang/test/Transforms/debug-module-1.f90]: Fixed.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 7]
```diff
@@ -99,6 +109,68 @@ void AddDebugInfoPass::handleDeclareOp(fir::cg::XDeclareOp declOp,
...
+  // it is used.
+  // FIXME: Currently we don't have the line number on which a module was
+  // declared. We are using a best guess of line - 1 where line is the source
+  // line of the first member of the module that we encounter.
+
+  if (result.second.modules.empty())
+    return;
+
+  scope = getOrCreateModuleAttr(result.second.modules[0], fileAttr, scope,
+                                line - 1, !globalOp.hasInitializationBody());
```
Comment (jeanPerier) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:162]: You should use `globalOp.isInitialized()` here and below. In some cases, the initial value is defined as an attribute, not with operations inside the body.

Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp]: Thanks for you suggestions. Changed now.

**Comment by kiranchandramohan** [comment]
> On a related note, while processing `DeclareOp` of such variables, are you aware of anything that will tell me that these variables have `save` attribute? I ask because I noticed that they are being added as local variable too.

I am not aware.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 8]
```diff
@@ -99,6 +109,68 @@ void AddDebugInfoPass::handleDeclareOp(fir::cg::XDeclareOp declOp,
...
+
+void AddDebugInfoPass::handleGlobalOp(fir::GlobalOp globalOp,
+                                      mlir::LLVM::DIFileAttr fileAttr,
+                                      mlir::LLVM::DIScopeAttr scope) {
+  mlir::ModuleOp module = getOperation();
+  mlir::MLIRContext *context = &getContext();
+  fir::DebugTypeGenerator typeGen(module);
+  mlir::OpBuilder builder(context);
+
+  auto result = fir::NameUniquer::deconstruct(globalOp.getSymName());
```
Comment (kiranchandramohan) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:139]: Nit: Could you spell auto?

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 9]
```diff
@@ -99,6 +109,68 @@ void AddDebugInfoPass::handleDeclareOp(fir::cg::XDeclareOp declOp,
...
+// The `module` does not have a first class representation in the `FIR`. We
+// extract information about it from the name of the identifiers and keep a
+// map to avoid duplication.
+mlir::LLVM::DIModuleAttr AddDebugInfoPass::getOrCreateModuleAttr(
+    const std::string &name, mlir::LLVM::DIFileAttr fileAttr,
+    mlir::LLVM::DIScopeAttr scope, unsigned line, bool decl) {
+  mlir::MLIRContext *context = &getContext();
+  mlir::LLVM::DIModuleAttr modAttr;
+  if (auto iter{moduleMap.find(name)}; iter != moduleMap.end())
+    modAttr = iter->getValue();
```
Comment (kiranchandramohan) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:121]: Nit: Braces here to match the else.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 10]
```diff
@@ -99,6 +109,68 @@ void AddDebugInfoPass::handleDeclareOp(fir::cg::XDeclareOp declOp,
...
+  else {
+    modAttr = mlir::LLVM::DIModuleAttr::get(
+        context, fileAttr, scope, mlir::StringAttr::get(context, name),
+        mlir::StringAttr(), mlir::StringAttr(), mlir::StringAttr(), line, decl);
```
Comment (kiranchandramohan) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:125]: Nit: comments here to spell the default attributes.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 11]
```diff
@@ -99,6 +109,68 @@ void AddDebugInfoPass::handleDeclareOp(fir::cg::XDeclareOp declOp,
...
+  // declared. We are using a best guess of line - 1 where line is the source
+  // line of the first member of the module that we encounter.
+
+  if (result.second.modules.empty())
+    return;
+
+  scope = getOrCreateModuleAttr(result.second.modules[0], fileAttr, scope,
+                                line - 1, !globalOp.isInitialized());
+
+  auto diType = typeGen.convertType(globalOp.getType(), fileAttr, scope,
```
Comment (kiranchandramohan) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:164]: Nit: Spell the type.

File: flang/test/Transforms/debug-module-2.f90 [hunk 12]
```diff
@@ -0,0 +1,39 @@
+! RUN: %flang_fc1 -emit-llvm -debug-info-kind=standalone %s -o - | FileCheck %s
+! RUN: %flang_fc1 -emit-llvm -debug-info-kind=line-tables-only %s -o - | FileCheck --check-prefix=LINEONLY %s
```
Comment (kiranchandramohan) [inline flang/test/Transforms/debug-module-2.f90:2]: LLVM IR tests should go to the Integration directory.

File: flang/test/Transforms/debug-module-1.f90 [hunk 13]
```diff
@@ -0,0 +1,39 @@
+! RUN: %flang_fc1 -emit-fir -debug-info-kind=standalone -mmlir --mlir-print-debuginfo %s -o - | \
+! RUN: fir-opt --cg-rewrite --mlir-print-debuginfo | fir-opt --add-debug-info --mlir-print-debuginfo | FileCheck %s
```
Comment (kiranchandramohan) [inline flang/test/Transforms/debug-module-1.f90:2]: Probably best to test the pass only.

**Review by kiranchandramohan — APPROVED** [review]
LG.

File: flang/test/Transforms/debug-module-2.f90 (hunk 12 above)
Comment (abidh) [reply flang/test/Transforms/debug-module-2.f90]: I have moved the testcase to Integration folder. There may be other IR testcase that I added and I will move them in separate PR.

File: flang/test/Transforms/debug-module-1.f90 (hunk 13 above)
Comment (abidh) [reply flang/test/Transforms/debug-module-1.f90:2]: The `AddDebugInfo` pass now operates on `XDeclareOp` so we need to run `CodeGenRewrite` before to convert the `DeclareOp`.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 9 above)
Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp]: Done.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 10 above)
Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp]: Done.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 11 above)
Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp]: Done.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 8 above)
Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp]: Done. Although I noticed that auto has been used in most places with `fir::NameUniquer::deconstruct`.

File: flang/test/Transforms/debug-module-1.f90 (hunk 13 above)
Comment (kiranchandramohan) [reply flang/test/Transforms/debug-module-1.f90:2]: I was trying to convey the following:
-> If it is the `AddDebugInfo` pass that you want to test, then give it the input after `cg-rewrite` and test its output.
-> If it is the `CodeGen` changes that you want to test, then give it the input before `Codegen` and test its output.

I am guessing the LLVM IR changes helps it easier to review and hence it might be OK to add it to the Integration directory. 

Generally tests that specifically checks the changes are easier to review and also maintain. This way of writing helps to ensure that when some unrelated changes happen these tests are not affected. For testing end to end flows we should ideally set up a separate buildbot or move these to the llvm-testsuite/Fortran/Debug.

Comment (abidh) [reply flang/test/Transforms/debug-module-1.f90:2]: I agree with the suggestion on using output of `cg-rewrite` to test the `AddDebugInfo`. My initial motivation for basing test on fortran source code was that it was more readable.

As action items,

I will update `AddDebugInfo` tests to be based on the output of `cg-rewrite`
I will add `codegen` tests when it is involved.
I will keep adding llvm ir tests in Integration for now.
